### PR TITLE
More Arm64 lowering/backend work

### PR DESCRIPTION
**Pull request by kddnewton**
* We now have encoding support for the LDR instruction for loading a PC-relative memory location
* You can now call add/adds/sub/subs with signed immediates, which switches appropriately based on sign
* We can now load immediates into registers appropriately, attempting to keep the minimal number of instructions:
  * If it fits into 16 bytes, we use just a single movz.
  * Else if it can be encoded into a bitmask immediate, we use a single mov.
  * Otherwise we use a movz, a movk, and then optionally another one or two movks.
* Fixed a bunch of code to do with the Op::Load opcode.
* We now handle GC-offsets properly for Op::Load by skipping around them with a jump instruction. (This will be made better by constant pools in the future.)
* Op::Lea is doing what it's supposed to do now.
* Fixed a bug in the backend tests to do with not using the result of an Op::Add.

---

Update since yesterday:

* Add the return register to the list of scratch registers (this is kind of a hidden requirement for the register allocation algorithm because it's going to purposefully grab that register).
* For CCalls add a split pass that will load all of the arguments into their appropriate register through a load instructions first in case they don't work with a mov.
* Similarly for CRet, do the same thing for the return register.
* Load the value for the Test instruction if it doesn't fit with one mov.
* When emitting conditional jumps and ccalls, be sure to load the addresses using the emit_load_value function in case they don't fit with one mov.
* Fixed one more test that had an add that wasn't using its return value.
* A couple of tests needed more registers allocated because of ARM.

### CODE REVIEW

File: yjit/src/backend/arm64/mod.rs [hunk 1]
```diff
@@ -61,10 +61,7 @@ impl Assembler
     /// Get the list of registers from which we can allocate on this platform
     pub fn get_alloc_regs() -> Vec<Reg>
     {
-        vec![
-            X12_REG,
-            X13_REG
-        ]
+        vec![C_RET_REG, X12_REG]
```
Comment (maximecb) [inline yjit/src/backend/arm64/mod.rs:64]: I don't really understand why you put the C return register here. Doesn't that risk other instructions getting allocated that register?

I didn't need to do that on the x86 side. The register allocation code (which should be platform-agnostic) should already automatically allocate the C return register to `ccall`.

File: yjit/src/backend/ir.rs [hunk 2]
```diff
@@ -534,71 +534,6 @@ impl Assembler
...
-                    if !val.special_const_p() {
-                        asm.load(opnd);
-                    }
-                }
-
-                opnd
-            }
-
-            opnds.into_iter().map(|opnd| map_opnd(opnd, asm)).collect()
-        }
```
Comment (maximecb) [inline yjit/src/backend/ir.rs:561]: Did you test the results of the changes to the x86 codegen? I would have preferred that this be part of a separate PR (or that I make this change on my end).

Comment (kddnewton) [reply yjit/src/backend/ir.rs:561]: We can revert the last commit if you'd like and pull it in later. Yes, I tested it against the x86 codegen, from what I could tell it generated the same code. All of the tests continued to pass as well.

Comment (kddnewton) [reply yjit/src/backend/ir.rs:561]: Sorry I didn't realize you wanted to write this code, I didn't mean to overstep here.

File: yjit/src/backend/arm64/mod.rs (hunk 1 above)
Comment (kddnewton) [reply yjit/src/backend/arm64/mod.rs:64]: We did do this on the X86 side. It's explicitly a part of the vec here: https://github.com/Shopify/ruby/blob/1f85630e00d1f4fb0744e6c5b94fe0e607244a2e/yjit/src/backend/x86_64/mod.rs#L74. This is necessary because we explicitly try to take that register from the pool here: https://github.com/Shopify/ruby/blob/1f85630e00d1f4fb0744e6c5b94fe0e607244a2e/yjit/src/backend/ir.rs#L684. 

I'm happy to change this if you want to, but I was trying to match what X86 did here.

File: yjit/src/backend/ir.rs (hunk 2 above)
Comment (maximecb) [reply yjit/src/backend/ir.rs:561]: I was able to test it locally. It works so it's all good.

Comment (maximecb) [reply yjit/src/backend/ir.rs:561]: This is making me realize that we need a custom test workflow for validating our progress with the new backend like we did during the Rust port though. I will get on that.

File: yjit/src/backend/arm64/mod.rs (hunk 1 above)
Comment (maximecb) [reply yjit/src/backend/arm64/mod.rs:64]: Ahhh right because we assume that call kills everything else that's live. Ok makes sense 👍
